src/main/java/teammates/ui/webapi/GetStudentsAction.java

```diff
@@ -42,34 +61,94 @@ void checkSpecificAccessControl() throws UnauthorizedAccessException {
-            return new JsonResult(new StudentsData(studentsForCourse));
-        } else if (teamName == null && hasSectionPrivilege) {
-            // request to get students by instructor with section privilege
-            List<StudentAttributes> studentsForCourse = logic.getStudentsForCourse(courseId);
-            List<StudentAttributes> studentsToReturn = new LinkedList<>();
-            Set<String> sectionsWithViewPrivileges = instructor.getSectionsWithPrivilege(privilegeName).keySet();
-            studentsForCourse.forEach(student -> {
-                if (sectionsWithViewPrivileges.contains(student.getSection())) {
-                    studentsToReturn.add(student);
-                }
-            });
-            return new JsonResult(new StudentsData(studentsToReturn));
+
+        if (isCourseMigrated(courseId)) {
+            Instructor instructor = sqlLogic.getInstructorByGoogleId(courseId, userInfo.id);
+            String privilegeName = Const.InstructorPermissions.CAN_VIEW_STUDENT_IN_SECTIONS;
+            boolean hasCoursePrivilege = instructor != null
+                    && instructor.isAllowedForPrivilege(privilegeName);
+            boolean hasSectionPrivilege = instructor != null
+                    && instructor.getSectionsWithPrivilege(privilegeName).size() != 0;
+
+            if (teamName == null && hasCoursePrivilege) {
+                // request to get all course students by instructor with course privilege
+                List<Student> studentsForCourse = sqlLogic.getStudentsForCourse(courseId);
+                StudentsData data = new StudentsData();
+                List<StudentData> studentDataList = studentsForCourse
+                        .stream()
+                        .map(StudentData::new)
+                        .collect(Collectors.toList());
+
+                data.setStudents(studentDataList);
+
+                return new JsonResult(data);
+            } else if (teamName == null && hasSectionPrivilege) {
+                // request to get students by instructor with section privilege
+                List<Student> studentsForCourse = sqlLogic.getStudentsForCourse(courseId);
+                List<Student> studentsToReturn = new LinkedList<>();
+                Set<String> sectionsWithViewPrivileges = instructor
+                        .getSectionsWithPrivilege(privilegeName).keySet();
+
+                studentsForCourse.forEach(student -> {
+                    if (sectionsWithViewPrivileges.contains(student.getTeam().getSection().getName())) {
+                        studentsToReturn.add(student);
+                    }
+                });
+
+                StudentsData data = new StudentsData();
+                List<StudentData> studentDataList = studentsToReturn
+                        .stream()
+                        .map(StudentData::new)
+                        .collect(Collectors.toList());
+
+                data.setStudents(studentDataList);
```

Comment:
Same here, the migrated entities should use the constructor. The datastore entities can use the workaround method.

Reply:
Sorry I miss that out..!